sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1744 @@
+
+    val hostSpillStorageSize = if (rapidsConf.offHeapLimitEnabled) {
+      // Disable the limit because it is handled by the RapidsHostMemoryStore
+      None
+    } else if (rapidsConf.hostSpillStorageSize == -1) {
+      // + 1 GiB by default to match backwards compatibility
+      Some(rapidsConf.pinnedPoolSize + (1024L * 1024 * 1024))
+    } else {
+      Some(rapidsConf.hostSpillStorageSize)
+    }
+    // this should hopefully be pinned, but it would work without
+    hostSpillBounceBuffer = HostMemoryBuffer.allocate(rapidsConf.spillToDiskBounceBufferSize)
+
+    chunkedPackBounceBufferPool = new DeviceBounceBufferPool {
+      private val bounceBuffer: DeviceBounceBuffer =
+        DeviceBounceBuffer(DeviceMemoryBuffer.allocate(rapidsConf.chunkedPackBounceBufferSize))
+      override def bufferSize: Long = rapidsConf.chunkedPackBounceBufferSize
+      override def nextBuffer(): DeviceBounceBuffer = {
+        // can block waiting for bounceBuffer to be released
+        bounceBuffer.acquire()
+      }
+      override def close(): Unit = {
+        // this closes the DeviceMemoryBuffer wrapped by the bounce buffer class
+        bounceBuffer.release()
+      }
+    }
+    storesInternal = new SpillableStores {
+      override var deviceStore: SpillableDeviceStore = new SpillableDeviceStore
+      override var hostStore: SpillableHostStore = new SpillableHostStore(hostSpillStorageSize)
+      override var diskStore: DiskHandleStore = new DiskHandleStore(conf)
+    }
+    val hostSpillStorageSizeStr = hostSpillStorageSize.map(sz => s"$sz B").getOrElse("unlimited")
+    logInfo(s"Initialized SpillFramework. Host spill store max size is: $hostSpillStorageSizeStr.")
+  }
+
+  def shutdown(): Unit = {
+    if (hostSpillBounceBuffer != null) {
+      hostSpillBounceBuffer.close()
+      hostSpillBounceBuffer = null
+    }
+    if (chunkedPackBounceBufferPool != null) {
+      chunkedPackBounceBufferPool.close()
+      chunkedPackBounceBufferPool = null
+    }
+    if (storesInternal != null) {
+      storesInternal.close()
+      storesInternal = null
+    }
+  }
+
+  def withHostSpillBounceBuffer[T](body: HostMemoryBuffer => T): T =
+    hostSpillBounceBuffer.synchronized {
+      body(hostSpillBounceBuffer)
+    }
+
+  var chunkedPackBounceBufferPool: DeviceBounceBufferPool = _
+
+  private[spill] def remove(handle: StoreHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    // so we use `storesInternal` directly.
+    handle match {
+      case ds: DeviceSpillableHandle[_] =>
+        synchronized {
+          Option(storesInternal).map(_.deviceStore)
+        }.foreach(_.remove(ds))
+      case hs: HostSpillableHandle[_] =>
+        synchronized {
+          Option(storesInternal).map(_.hostStore)
+        }.foreach(_.remove(hs))
+      case dh: DiskHandle =>
+        synchronized {
+          Option(storesInternal).map(_.diskStore)
+        }.foreach(_.remove(dh))
```

Comment:
We shouldn't have to compute at runtime how to match a handle with a store. Handles should know what store they are using (the spill framework explicitly advertises a limited set of stores).  All of the uses of this method are called in contexts where it's known at compile time what type of handle is trying to be removed from a store.

Reply:
ok, I can split that back up.

Reply:
Here you go: https://github.com/NVIDIA/spark-rapids/pull/11747/commits/88c18b73e25f87e9dbfd2dbfb63f9b218c340365